infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/SecureServerExposerFactory.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.kubernetes.server.secure;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import org.eclipse.che.api.core.model.workspace.runtime.RuntimeIdentity;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
+import org.eclipse.che.workspace.infrastructure.kubernetes.server.external.ExternalServerExposerStrategy;
+import org.eclipse.che.workspace.infrastructure.kubernetes.server.secure.SecureServerExposer.DefaultSecureServerExposer;
+import org.eclipse.che.workspace.infrastructure.kubernetes.server.secure.jwtproxy.JwtProxySecureServerExposerFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** @author Sergii Leshchenko */
+public class SecureServerExposerFactory<T extends KubernetesEnvironment> {
+  private static final Logger LOG = LoggerFactory.getLogger(SecureServerExposerFactory.class);
+
+  private final boolean agentsAuthEnabled;
+  private final String serverExposer;
+
+  private final ExternalServerExposerStrategy<T> exposerStrategy;
+  private final JwtProxySecureServerExposerFactory<T> jwtProxySecureServerExposerFactory;
+
+  @Inject
+  public SecureServerExposerFactory(
+      @Named("che.agents.auth_enabled") boolean agentsAuthEnabled,
```

Comment:
Thank, Good point! It makes more sense to use `servers` word there instead of `agents`. I will change it.

Reply:
@garagatyi Do you think it may be `che.server.secure_exposer` property which is Che System scope or it would be better to make it K8s/OS scope like `che.infra.kubernetes.server_secure_exposer`?

Reply:
interesting question) On one hand it would be more complex to configure Che if we had several properties for several infras. On the other hand, we don't have anything like this for the Docker infra. 
I think it is not so critical, and I would opt for the 1st option.

Reply:
OK, I will rename it to `che.server.secure_exposer` and we'll rename it to infraspecific if it will be needed to have different authentication proxy backends for different infrastructures.

Reply:
I'm OK to leave the old name. I missed that it is old.

Reply:
`che.agents.auth_enabled` leaved as it is and `che.agents.secure_server_exposer` is renamed to `che.server.secure_exposer`